src/networking.c

```diff
@@ -1631,6 +1632,12 @@ void freeClient(client *c) {
     reqresReset(c, 1);
 #endif
 
+    /* Remove the contribution that this client gave to our
+     * incrementally computed memory usage. */
+    if (c->conn)
+        server.stat_clients_type_memory[c->last_memory_type] -=
+            c->last_memory_usage;
```

Comment:
start version: 7.0.0
introduced: #8687
reason: touch `server.stat_clients_type_memory` without GIL

```
WARNING: ThreadSanitizer: data race (pid=90167)
  Write of size 8 at 0x0001006f41e0 by main thread (mutexes: write M0, write M1, write M2, write M3):
    #0 updateClientMemoryUsage server.c:956 (redis-server:arm64+0x10001f760)
    #1 clientsCron server.c:1116 (redis-server:arm64+0x1000201fc)
    #2 serverCron server.c:1451 (redis-server:arm64+0x100021b10)
    #3 processTimeEvents ae.c:331 (redis-server:arm64+0x1000100b8)
    #4 aeProcessEvents ae.c:466 (redis-server:arm64+0x10000f614)
    #5 aeMain ae.c:496 (redis-server:arm64+0x1000103bc)
    #6 main server.c:7212 (redis-server:arm64+0x10003e76c)

  Previous write of size 8 at 0x0001006f41e0 by thread T7:
    #0 freeClient networking.c:1684 (redis-server:arm64+0x10005f30c)
    #1 moduleFreeContext module.c:834 (redis-server:arm64+0x1001821ac)
    #2 RM_FreeThreadSafeContext module.c:8494 (redis-server:arm64+0x10019dcf8)
    #3 worker <null>:17631300 (blockedclient.so:arm64+0x828)
```

Reply:
we can't / shouldn't evict them anyway.
if we tracked them, it was wrong to do that.
we can list this as a fix (not about thread race) in the top comment.

Reply:
I've never been able to remember how I reproduced it.
I remember using RM_Call to reproduce it, but I forget which command, not the following `client no-evict` command.

Config:
```
maxmemory-clients 1g
```

Command:
```c
RedisModule_Call(ctx,"client","cc","no-evict","off");
```

Patch:
```c
int clientEvictionAllowed(client *c) {
    serverAssert(c->conn);
    if (server.maxmemory_clients == 0 || c->flags & CLIENT_NO_EVICT) {
        return 0;
    }
    int type = getClientType(c);
    return (type == CLIENT_TYPE_NORMAL || type == CLIENT_TYPE_PUBSUB);
}
```

`serverAssert(c->conn);` will be triggered.

Reply:
please als have a look at https://github.com/redis/redis/pull/12817#issuecomment-1851899163 and top comment(7).

Reply:
i don't think using RM_Call to call the CLIENT command is valid.
specifically the ones manipulating the current client, like enabling tracking, and so on.
the user may be wanting to operate on the calling client, not the fake client, but that's not currently supported, and i think we should just disallow or disregard this case.